### PR TITLE
[Arbitrary txs] Improve function data decoding

### LINKED ISSUE #3875: [Arbitrary Txs] Improve function data decoding
## Description

The current `decodeArbitraryFunction` utility added in [block-ingestor#299](https://github.com/JoinColony/block-ingestor/pull/299) works in most cases but is susceptible to hash collisions between different function signatures.

For example, consider the `mint(address,uint256)` function with hash `0x40c10f19`. This hash is shared by `cat642998653(address,uint256)` (https://www.4byte.directory/signatures/?bytes4_signature=0x40c10f19), and without access to the original contract ABI, we cannot distinguish between these functions.

~~There are a few competing solutions but one discussed with @rdig and @Nortsova involves block-ingestor following these steps when it attempts to decode an arbitrary function:~~
~~1. Call Arbiscan API (just like in #3694) to try and fetch the original ABI. This will only work if the contract source code has been verified.~~
~~2. If the above fails, use existing `decodeArbitraryFunction` that calls 4byte.directory. This function needs to be modified to return ALL possible matches, not just the first one it can decode.~~

**Edit:** We have since found that there could also be collisions between functions which arguments do not match, for example `approve(address,uint256)` has the same signature as `watch_tg_invmru_2f69f1b(address,address)` and decoding would 'succeed' for both, making it impossible to tell which was originally called.

To handle this case, we will instead store the original version of the ABI used to create arbitrary transaction as action metadata, and use that in the UI to decode functions on the fly.

## Acceptance criteria
- [ ] Original ABIs are stored in JSON format in the action metadata
- [ ] The UI performs decoding on the fly

### PR DESCRIPTION
**Pull request by jakubcolony**
## Description

Don't be put off by the number of files changed, 80% of them are part of a small function arguments refactor! 

See #3875 for context.

This PR shifts the function data decoding to the UI and makes it rely on the original ABI used to create the arbitrary transaction.

[block-ingestor PR](https://github.com/JoinColony/block-ingestor/pull/305)

## Testing

To test the decoding without hacking a sandbox to call the util, we'll deploy a test contract that has functions with a variety of argument types we expect to encounter.

1. Ensure your dev env is fully running.
2. Create a new file called `TestContract.sol` in your CDapp root directory with the following contents:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract TestContract {
    function simpleFunction(uint256 uint256Arg) public {}

    function anotherSimpleFunction(
        string memory stringArg,
        address addressArg
    ) public {}

    function testFunction(
        address addressArg,
        string memory stringArg,
        bool boolArg,
        uint8 uint8Arg,
        uint256 uint256Arg,
        bytes memory bytesArg
    ) public {}
}
```
4. Run the following in your terminal to copy the contract over to the network container and compile it:
```bash
docker cp TestContract.sol network:/colonyCDappBackend/colonyNetwork/contracts && docker exec -it -w /colonyCDappBackend/colonyNetwork network npx hardhat compile
```
5. Start the hardhat console with `npm run hardhat`, then run the following code line-by-line:
```js
cf = await ethers.getContractFactory('TestContract')
c = await cf.deploy()
c.address
```
The last expression will print the address of the newly deployed contract. Copy it for later use.

In the UI:
1. Create a custom transaction action (under Manage colony).
2. In the Add transaction modal, paste the address of the deployed contract.
3. Because there's no API to get ABI for local contracts, paste it in the textarea manually:
```json
[
  {
    "inputs": [
      {
        "internalType": "string",
        "name": "stringArg",
        "type": "string"
      },
      {
        "internalType": "address",
        "name": "addressArg",
        "type": "address"
      }
    ],
    "name": "anotherSimpleFunction",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "uint256Arg",
        "type": "uint256"
      }
    ],
    "name": "simpleFunction",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "address",
        "name": "addressArg",
        "type": "address"
      },
      {
        "internalType": "string",
        "name": "stringArg",
        "type": "string"
      },
      {
        "internalType": "bool",
        "name": "boolArg",
        "type": "bool"
      },
      {
        "internalType": "uint8",
        "name": "uint8Arg",
        "type": "uint8"
      },
      {
        "internalType": "uint256",
        "name": "uint256Arg",
        "type": "uint256"
      },
      {
        "internalType": "bytes",
        "name": "bytesArg",
        "type": "bytes"
      }
    ],
    "name": "testFunction",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  }
]
```
4. Select one of the available functions and fill its arguments with data of your choice (the validation should help you out!).
6. Add more transactions or create an action with just one.

You will be taken to a rudimentary UI showing the details of the arbitrary transactions your action contained:
- method name
- argument names with their types 
- values of the arguments

image

Resolves #3875

### CODE REVIEW

**Review by bassgeta — APPROVED** [review]
Nice one, makes sense to do the heavy lifting on the client, and if it solves our collision issues, even better!
Hardhat deployed the test contract :heavy_check_mark: 
image
Filled out the UI and then got to the completed action view :heavy_check_mark: 
image
image
image
Creating an action with 5 transactions also worked and didn't break :heavy_check_mark: 
image

**Review by Nortsova — APPROVED** [review]
Nice work @jakubcolony ✨  Thanks for this changes 🙌 

You found a great solution for this issue 👏 

image

**Review by mmioana — APPROVED** [review]
Really nice work on this one @jakubcolony 🙌 

Followed your testing steps and can confirm it works flawlessly 💯 

Copied the custom contract to the network container and compiled it
Screenshot 2024-12-18 at 23 58 14

Deployed the custom contract
Screenshot 2024-12-18 at 23 58 54

Added a transaction to the `Custom transactions` action
Screenshot 2024-12-19 at 00 00 44

And another one
Screenshot 2024-12-19 at 00 01 22

The transactions are correctly listed before submitting the form
Screenshot 2024-12-19 at 00 01 56

And the action got completed ✅ 
Screenshot 2024-12-19 at 00 02 17

Nice work!!!

File: src/redux/sagas/arbitraryTxs/arbitraryTx.ts [hunk 1]
```diff
@@ -106,7 +103,27 @@ function* arbitraryTxSaga({
       },
     } = yield waitForTxResult(makeArbitraryTransactions.channel);
 
-    yield createActionMetadataInDB(txHash, customActionTitle);
+    const abisByAddress = transactions.reduce<Record<string, string>>(
+      (acc, transaction) => {
+        const checksummedAddress = utils.getAddress(
+          transaction.contractAddress,
+        );
```
[the remark this answers is not in the document]
Comment (jakubcolony) [reply src/redux/sagas/arbitraryTxs/arbitraryTx.ts:110]: The form validation should prevent invalid addresses from being passed to the saga. But, if one still makes it through, the saga should throw an error and prevent any transactions from being sent.

**Pull request merged**